graphql/src/subscription/mod.rs

```diff
@@ -228,9 +222,16 @@ where
     };
 
     // We have established that this exists earlier in the subscription execution
-    let subscription_type = sast::get_root_subscription_type(&ctx.schema.document).unwrap();
-
-    let result = execute_selection_set(&ctx, selection_set, subscription_type, &None);
+    let subscription_type = sast::get_root_subscription_type(&ctx.schema.document)
+        .unwrap()
+        .clone();
+
+    let result = graph::spawn_blocking_allow_panic(async move {
+        execute_selection_set(&ctx, &selection_set, &subscription_type, &None)
+    })
+    .await
```

Comment:
This should helps since now the number of blocking threads we need is no longer dependent on the number of subscriptions, but we can still exhaust the blocking pool because of the thundering herd behavior of subscriptions. If the individual queries for each subscription take a while to run (say 1s which is not that hard to cause) the first few subscriptions through here will exhaust the connection pool, which will cause following subscriptions to wait for those queries to finish, eventually filling up the blocking pool. We should gate spawning the blocking thread on a semaphore that is sized so that we do not exhaust the connection pool (say allows 75% of the connection pool through)

Reply:
What we should really do is change the store to move actual work to the blocking pool itself by having some internal function `Store.with_connection(f:Fn(Connection)->Result)` which first acquires a semaphore sized according to the max number of connections in the pool and then executes `f` on the blocking pool. We'd then change all `Store` methods that right now just get a connection to use `with_connection` and do their work inside that.

Reply:
It does make sense to throttle subscription queries so they don't delay normal queries. Ideally we'd have query nodes dedicated only to subscriptions. I'll put an async semaphore here as you suggest.